src/workerd/api/actor-state.c++

```diff
@@ -713,6 +715,62 @@ jsg::Promise<jsg::Value> DurableObjectState::blockConcurrencyWhile(jsg::Lock& js
   return IoContext::current().blockConcurrencyWhile(js, kj::mv(callback));
 }
 
+void DurableObjectState::acceptWebSocket(
+    jsg::Ref<WebSocket> ws,
+    jsg::Optional<kj::Array<kj::String>> tags) {
+  JSG_ASSERT(!ws->isAccepted(), Error,
+      "Cannot call `acceptWebSocket()` if the WebSocket was already accepted via `accept()`");
+  JSG_ASSERT(ws->LOCAL == WebSocket::Locality::LOCAL, Error,
+      "Outgoing WebSocket connections cannot be hibernated.");
+
+  // We need to get a HibernationManager to give the websocket to.
+  auto& a = KJ_REQUIRE_NONNULL(IoContext::current().getActor());
+  if (a.getHibernationManager() == nullptr) {
+    a.setHibernationManager(
+        kj::refcounted<HibernationManagerImpl>(
+            a.getLoopback(), KJ_REQUIRE_NONNULL(a.getHibernationEventType())));
+  }
+  // HibernationManager's acceptWebSocket() will throw if the websocket is in an incompatible state.
+  // Note that not providing a tag is equivalent to providing an empty tag array.
+  // Any duplicate tags will be ignored.
+  kj::Array<kj::String> distinctTags = [&]() -> kj::Array<kj::String> {
+    KJ_IF_MAYBE(t, tags) {
+      kj::HashSet<kj::String> seen;
+      size_t tagCount = 0;
+      for (auto tag = t->begin(); tag < t->end(); tag++, tagCount++) {
```

Comment:
That's about the idea I suggested, but isn't this counting even repeated tags? Maybe you should increment only inside `if (!seen.contains(*tag)) `. Or is the idea to fail even if there are only < 10 distinct tags?

Reply:
Thanks for catching that, I was too sleepy yesterday :upside_down_face: